### ISSUE TITLE
Modify index.html with ngx-build-plus

### ISSUE THREAD

**Issue by Timebutt** [manfredsteyer/ngx-build-plus#140]
Hi,

I am upgrading an existing `Angular 7` project to the latest `Angular 8.2.10` and noticed one of the plugins we run using `ngx-build-plus` to add a `window.__env` variable to our `index.html` file doesn't work anymore.

This is the code for it:

```
class EnvPlugin {
    constructor(env) {
        this._options = { env };
    }

    apply(compiler) {
        compiler.hooks.emit.tapPromise('env-webpack-plugin', async compilation => {
            if (compilation.assets['index.html']) {
                const indexSource = compilation.assets['index.html'];
                const envInsertionPoint = indexSource
                    .original()
                    .source()
                    .indexOf('</head>');

                indexSource.insert(
                    envInsertionPoint,
                    `<script>window.__env = ${JSON.stringify(this._options.env || {})}</script>`
                );
            }
        });
    }
}
```

After some digging, I found out that Angular 8 CLI introduced a breaking API change that leads to `index.html` no longer being added to the `compilation.assets` array in the above example (see [this article](https://blog.angularindepth.com/customizing-angular-cli-build-an-alternative-to-ng-eject-v2-c655768b48cc)). 

In the above article, they suggest using `@angular-builders/custom-webpack` and using their `indexTransform` option to transform the `index.html`. It requires switching the `angular.json` builder to theirs however, and I want to keep using `ngx-build-plus` for my other plugins.

Is there a way I can still run a transformation as demonstrated above on my CLI generated `index.html` using `ngx-build-plus`, and if so: how do I go about that exactly?

Thanks in advance!
Nils

**Comment by manfredsteyer** [comment]
Hi Timebutt,

thanks for reaching out. I would say, both plugins do basically the same. The underlying issue here is indeed that the way the CLI generates the `index.html` has changed since Angular 8. Before, webpack was responsible for generating the `index.html`; now it is done by the builder to address differential loading.

To solve this issue with `ngx-build-plus`, you just needed a `ngx-build-plus` plugin, which is basically an object with up to three methods [1].

Just can put your logic for changing the `index.html` into the post method which is executed after the build. I would use Node's `fs`` object for this.

You can point to your plugin using the `--plugin` command line switch or using the respective settings within `angular.json`.

Best wishes,
Manfred

[1] https://github.com/manfredsteyer/ngx-build-plus#using-plugins

**Comment by Timebutt** [comment]
Hey Manfred, thanks for your response. We are currently looking into how we can fix our current situation. Your proposal still poses issues for a dev build in the way I understand correctly. We're currently exploring the `Architect API` to see if we can use a tailored solution. I'll close this for now, as the issue is unrelated to `ngx-build-plus`.

Thanks again!

**Comment by klaascuvelier** [comment]
Hey @manfredsteyer, thanks for your quick reply and suggestion.

I work together with @Timebutt and we investigated the idea of using an ngx-build-plus plugin. We tried a combination of plugin + web app config and just a plugin without extra web pack config but found out we were not able to make it work for both `ng serve` and `ng build`.

After some investigation into your builder and some other open-source builders, we decided to implement our own. As angular-devkit provides an easy way to create a builder and add custom transforms it was fairly straightforward to get our use case up and running.

Additionally by creating our own builder, it does not have to be a very generic solution which allows any use case (like your or other open-source builders have to be). We can _hide_ our transforms right inside the builder itself, which makes our actual commands much simpler.
We can just do `ng serve` or `ng build` without the extra `--plugin` or `--extra-webpack-config` options. This improves the developer experience for our team as well.

Once more thanks for the help, and thank you for your open-source work for the community :bow:
